File: storage/db-core/src/main/java/com/coffee/gu/coupon/IssuedCouponEntity.java
package com.coffee.gu.coupon;

import com.coffee.gu.Principal;
import com.coffee.gu.enums.IssuedCouponState;
import com.coffee.gu.BaseEntity;
import com.coffee.gu.enums.PrincipalType;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Table;

@Table(name = "issued_coupon")
@Entity
public class IssuedCouponEntity extends BaseEntity {

    private String principalKey;
    @Enumerated(EnumType.STRING)
    private PrincipalType principalType;
    private Long couponId;
    private IssuedCouponState state;

    protected IssuedCouponEntity() {}

    public IssuedCouponEntity(String principalKey, PrincipalType principalType, Long couponId, IssuedCouponState state) {
        this.principalKey = principalKey;
        this.principalType = principalType;
        this.couponId = couponId;
        this.state = state;
    }

    public static IssuedCouponEntity from(IssuedCoupon issuedCoupon) {
        return new IssuedCouponEntity(
                issuedCoupon.getPrincipal().getKey(),
                issuedCoupon.getPrincipal().getType(),
                issuedCoupon.getCoupon().getId(),
                issuedCoupon.getState()
        );
    }

    public IssuedCoupon toModel(Coupon coupon) {
        return new IssuedCoupon(
                this.id,
                new Principal(this.principalKey, this.principalType),
                this.state,
                coupon
        );
    }

    public String getPrincipalKey() {
        return principalKey;
    }

    public PrincipalType getPrincipalType() {
        return principalType;
    }

    public Long getCouponId() {
        return couponId;
    }

    public IssuedCouponState getState() {
        return state;
    }
}

File: storage/db-core/src/main/kotlin/com/coffee/gu/coupon/IssuedCouponEntity.kt
package com.coffee.gu.coupon

import com.coffee.gu.BaseEntity
import com.coffee.gu.Principal
import com.coffee.gu.enums.IssuedCouponState
import com.coffee.gu.enums.PrincipalType
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.Table

@Table(name = "issued_coupon")
@Entity
class IssuedCouponEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0,
    val principalKey: String,
    @Enumerated(EnumType.STRING)
    val principalType: PrincipalType,
    val couponId: Long,
    @Enumerated(EnumType.STRING)
    val state: IssuedCouponState,
) : BaseEntity() {

    fun toModel(coupon: Coupon): IssuedCoupon {
        return IssuedCoupon(
            id = id,
            principal = Principal(principalKey, principalType),
            state = state,
            coupon = coupon,
        )
    }

    companion object {
        fun from(issuedCoupon: IssuedCoupon): IssuedCouponEntity = IssuedCouponEntity(
            id = issuedCoupon.id,
            principalKey = issuedCoupon.principal.key,
            principalType = issuedCoupon.principal.type,
            couponId = issuedCoupon.coupon.id ?: 0L,
            state = issuedCoupon.state,
        )
    }
}
